File: retos-ts/22_combinaciones.ts
// Resultado 5 ⭐
function generateGiftSets(gifts: string[]): string[][] {
    const result: string[][] = []

    function getCombinations(sI: number, currSet: string[]) {
        for (let i = sI; i < gifts.length; i++) {
            currSet.push(gifts[i])
            result.push([...currSet])
            getCombinations(i + 1, currSet)
            currSet.pop()
        }
    }

    getCombinations(0, [])

    return result.sort((a, b) => a.length - b.length)
}

File: retos-python/22_combinaciones.py
# Resultado 5 ⭐
def generate_gift_sets(gifts):
    result = []

    def backtrack(start_index, current_set):
        for i in range(start_index, len(gifts)):
            current_set.append(gifts[i])
            result.append(current_set[:])
            backtrack(i + 1, current_set)
            current_set.pop()

    backtrack(0, [])
    return sorted(result, key=len)
